File: routes/proxy.go
package routes

import (
	"encoding/json"
	"io"
	"log"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
)

func RegisterProxyRoutes(r *gin.Engine) {
	r.GET("/proxy", func(c *gin.Context) {
		startAt := c.Query("startAt")
		endAt := c.Query("endAt")

		umamiURL := os.Getenv("UMAMI_URL")
		websiteID := os.Getenv("UMAMI_WEBSITE_ID")
		token := os.Getenv("UMAMI_TOKEN")

		url := umamiURL + "/api/websites/" + websiteID + "/stats/?startAt=" + startAt + "&endAt=" + endAt

		req, err := http.NewRequest("GET", url, nil)
		if err != nil {
			log.Println("Proxy request creation failed:", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "proxy failed"})
			return
		}

		req.Header.Set("Authorization", "Bearer "+token)
		req.Header.Set("User-Agent", "umami-proxy/1.0")
		req.Header.Set("Accept", "application/json")

		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			log.Println("Proxy fetch error:", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "proxy failed"})
			return
		}
		defer resp.Body.Close()

		contentType := resp.Header.Get("Content-Type")
		body, _ := io.ReadAll(resp.Body)

		if resp.StatusCode != http.StatusOK || contentType == "" || !isJSON(contentType) {
			log.Printf("Invalid response from Umami [%d] [%s]: %s\n", resp.StatusCode, contentType, string(body))
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Invalid response from Umami"})
			return
		}

		var data map[string]interface{}
		if err := json.Unmarshal(body, &data); err != nil {
			log.Println("JSON parse failed:", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "parse failed"})
			return
		}

		c.JSON(http.StatusOK, data)
	})
}

func isJSON(contentType string) bool {
	return contentType == "application/json" || len(contentType) >= 16 && contentType[:16] == "application/json"
}

File: routes/proxy.js
import express from 'express';
const router = express.Router();

const WEBSITE_ID = process.env.UMAMI_WEBSITE_ID;
const UMAMI_URL = process.env.UMAMI_URL;
const UMAMI_TOKEN = process.env.UMAMI_TOKEN;

router.get('/proxy', async (req, res) => {
  const { startAt, endAt } = req.query;
  const url = `${UMAMI_URL}/api/websites/${WEBSITE_ID}/stats/?startAt=${startAt}&endAt=${endAt}`;

  try {
    const response = await fetch(url, {
      headers: {
        Authorization: `Bearer ${UMAMI_TOKEN}`,
        'User-Agent': 'umami-proxy/1.0',
        Accept: 'application/json',
      },
    });

    const contentType = response.headers.get('content-type') || '';
    const raw = await response.text();

    if (!response.ok || !contentType.includes('application/json')) {
      return res.status(500).json({ error: 'Invalid response from Umami' });
    }

    const data = JSON.parse(raw);
    res.json(data);
  } catch (err) {
    console.error('Proxy fetch error:', err);
    res.status(500).json({ error: 'proxy failed' });
  }
});

export default router;
